### PR TITLE
Read password hash before fork

### PR DESCRIPTION
**Pull request by sertonix**
This ensures that the parent properly errors only if the password cannot be read.

### CODE REVIEW

**Comment by sertonix** [comment]
This might introduce more issues similar to !175. I think a lot more issues would be fixed by removing the safety check.

File: shadow.c [hunk 1]
```diff
@@ -15,11 +16,15 @@
 #include "swaylock.h"
 
 void initialize_pw_backend(int argc, char **argv) {
-	if (geteuid() != 0) {
+	int fd_shadow = open("/etc/shadow", O_CLOEXEC | O_RDONLY);
```
Comment (kennylevinsen) [inline shadow.c:19]: If you do not want a file descriptor, consider using `access(2)` instead to just check the file permissions.

**Comment by kennylevinsen** [comment]
This will not work for #175, where the shadow file is located elsewhere. For that, the checks would have to be replaced with a password retrieval attempt (the start of `run_pw_backend_child`).

**Comment by sertonix** [comment]
> This will not work for #175, where the shadow file is located elsewhere. For that, the checks would have to be replaced with a password retrieval attempt (the start of `run_pw_backend_child`).

I looked at the code and found a better solution. Will push as soon as it is done.

**Comment by sertonix** [comment]
Is it safe to pass a pointer to a fork? This code does so.

**Comment by emersion** [comment]
A downside of this approach is that changing the user's password while swaylock is running won't work. In other words, if someone forgets their password and tries to change it via SSH, swaylock still won't unlock. Maybe that's not such a big deal since it's possible to send a signal to swaylock to unlock itself via SSH too.

**Comment by sertonix** [comment]
> A downside of this approach is that changing the user's password while swaylock is running won't work.

I am pretty sure that is already how it works even without my change.

**Comment by emersion** [comment]
Hm, indeed!

**Comment by sertonix** [comment]
The privilege drop would also not work with reading the password later.

**Comment by sertonix** [comment]
@emersion What is the blocker for this PR?

**Review by emersion — APPROVED** [review]
Thanks!

**Pull request merged**